### DISCUSSION TITLE
Error with inbound emails

### DISCUSSION THREAD

**Discussion by AhmedNadar** [DavidColby/hotwired_ats_code#22] [General]
Hey everyone,

In chapter 7 [Building a notification system](https://book.hotwiringrails.com/chapters/7 ), I'm facing an error regards sending inbound emails with ActionMailbox using [Rails conductor form](http://localhost:3000/rails/conductor/action_mailbox/inbound_emails/new). 

When I send an email, it is failed, 
<img width="320" alt="image" src="https://user-images.githubusercontent.com/174362/196047945-b8658283-b178-4137-9600-6efe9aafd862.png">

... but still, it's sent as I see an inbound emails list on the applicant show page.
<img width="409" alt="image" src="https://user-images.githubusercontent.com/174362/196047973-efd9a6a0-6917-42af-9eb2-51d03e490e24.png">

While the app log shows:
```shell
@uri=#<URI::GID gid://hotwired-ats/ActionMailbox::InboundEmail/9872c7e1-6b83-4420-adf3-0bc5297a40e2>>

12:37:23 web.1  | [ActiveJob] [ActionMailbox::RoutingJob] [d97269ca-b6b7-4494-bbed-0160db6bfa99] Error performing 
ActionMailbox::RoutingJob (Job ID: d97269ca-b6b7-4494-bbed-0160db6bfa99) from Async(default) in 535.09ms:
Psych::DisallowedClass (Tried to load unspecified class: Applicant):

12:37:23 web.1  | /Users/ahmed/Projects/Ruby/hotwired_ats/app/models/email.rb:21:in `create_notification'

12:37:23 web.1  | /Users/ahmed/Projects/Ruby/hotwired_ats/app/mailboxes/applicant_replies_mailbox.rb:10:in `process'
```
It looks like `Psych` library has an issue with `Applicant`. Looking into `create_notification` method with `email.rb` class.
Searching online i found [there is a possible escalation to RCE when using YAML serialized columns in Active Record](https://discuss.rubyonrails.org/t/cve-2022-32224-possible-rce-escalation-bug-with-serialized-columns-in-active-record/81017).

What I understand is, we are serializing `params` in [`notification.rb`](https://github.com/DavidColby/hotwired_ats_code/blob/f3e8c55a2516a84b71a3a332783fe0f2a1b37b48/app/models/notification.rb#L8). Here we didn't specify [`class_name_or_coder`](https://api.rubyonrails.org/classes/ActiveRecord/AttributeMethods/Serialization/ClassMethods.html#method-i-serialize) therefore default is `YAML`. 
> default - The attribute value will be serialized as YAML. The attribute value must respond to to_yaml.

But `params` data_type is [`:jsonb`](https://github.com/DavidColby/hotwired_ats_code/blob/f3e8c55a2516a84b71a3a332783fe0f2a1b37b48/db/migrate/20220123155450_create_notifications.rb#L6). 

When I tried to specify `class_name_or_coder` as JSON to [avoid security impact](https://discuss.rubyonrails.org/t/cve-2022-32224-possible-rce-escalation-bug-with-serialized-columns-in-active-record/81017#impact-1), `serialize :params, JSON` it throw an error:

```shell
Column `params` of type ActiveRecord::ConnectionAdapters::PostgreSQL::OID::Jsonb does not support `serialize` feature.
Usually it means that you are trying to use `serialize`
on a column that already implements serialization natively.
```

Has anyone faced such an issue? Any feasible workarounds to solve this usse?
Thanks

**Comment by ZilvinasKucinskas** [comment]
Rails 7.0.3.1 introduced a [security patch](https://discuss.rubyonrails.org/t/cve-2022-32224-possible-rce-escalation-bug-with-serialized-columns-in-active-record/81017) that breaks the serialization. 

For the purpose of this tutorial, the simplest workaround would be to use:

```ruby
config.active_record.use_yaml_unsafe_load = true
```

in the `application.rb`.
